Integrate SoI in the main loop

anwu1219 opened the pull request:
1. This PR modularized the main loop to improve its readability. This includes calling a handleSatisfyingAssignmentToConvexRelaxation() method when all the linear constraints are satisfied. That method will invoke either the Reluplex or the SoI logic depending on the options.
2. Introduces the performLocalSearch() method in the Engine.cpp which mainly invokes methods in SoIManager to carry out the local search.
3. Introduces a new class, PseudoImpactTracker, which maintains a heap structure of the PseudoImpact as a branching heuristics.
4. Added the PseudoImpact branching heuristics and did a small refactoring of the SmtCore for code readability.

anwu1219 commented on src/engine/CDSmtCore.cpp:
```diff
@@ -2,7 +2,7 @@
 /*! \file CDSmtCore.cpp
```
There are no unit tests for this class. I did my best effort to integrate the SoI change to this. But we probably need to revisit testing the correctness of this class later.

guykatzz commented on src/configuration/GlobalConfiguration.cpp:
```diff
@@ -18,6 +18,14 @@
+
+// Whether to use SoI instead of Reluplex for local search for satisfying assignments
+//to non-linear constraint.
+const bool GlobalConfiguration::USE_DEEPSOI_LOCAL_SEARCH = false;
```
why not set this to true by default?

anwu1219 replied:
Will do after extending each PLConstraint with SoI as discussed off-line

guykatzz commented on src/engine/PLConstraintScoreTracker.h:
```diff
@@ -0,0 +1,109 @@
+        , _score( score )
+    {};
+
+    bool operator<(const ScoreEntry& other ) const
```
space before const, space before & and not after

anwu1219 replied:
whoops, done.

anwu1219 commented on src/engine/Engine.cpp:
```diff
@@ -416,6 +358,132 @@ void Engine::mainLoopStatistics()
+        performMILPSolverBoundedTighteningForSingleLayer
+            ( _networkLevelReasoner->getLayerIndexToLayer().size() - 1 );
+}
+
```
done

anwu1219 commented on src/engine/Engine.cpp:
```diff
@@ -416,6 +358,132 @@ void Engine::mainLoopStatistics()
+}
+
+
+inline bool Engine::handleSatisfyingAssignmentToConvexRelaxation()
```
This method is called when a satisfying assignment is already found for the linear constraints. The next step is to attempt to satisfy the PLConstraints as well. I renamed it to handleSatisfyingAssignmentToLinearConstraints()

guykatzz replied:
I still think it could be a bit misleading, especially because the linear constraints are already satisfied. Perhaps "adjustSatisfyingAssigmentToNonLinearConstraints", or something similar? I know it's a bit cluttered...

anwu1219 replied:
I see now how it can be confusing.. Probably adjustAssignmentToSatisfyNonLinearConstraints" is the least confusing. Will change to that.

anwu1219 commented on src/engine/Engine.cpp:
```diff
@@ -416,6 +358,132 @@ void Engine::mainLoopStatistics()
+    }
+    else
+    {
+        return performLocalSearch();
```
done

anwu1219 commented on src/engine/Engine.cpp:
```diff
@@ -2514,6 +2633,87 @@ const Preprocessor *Engine::getPreprocessor()
+        // We have the "local" effect of change the cost term of some
+        // PLConstraints in the phase pattern. Use this information to influence
+        // the branching decision.
+        updatePseudoImpact( costOfLastAcceptedPhasePattern,
```
Renamed to updatePseudoImpactWithSoICosts()

anwu1219 commented on src/engine/Engine.cpp:
```diff
@@ -2524,53 +2724,28 @@ void Engine::minimizeHeuristicCost( const LinearExpression &heuristicCost )
     bool localOptimaReached = false;
     while ( !localOptimaReached )
```
Yep.. Done

anwu1219 commented on src/engine/Engine.cpp:
```diff
@@ -2514,6 +2633,87 @@ const Preprocessor *Engine::getPreprocessor()
+            if ( allPlConstraintsHold() )
+            {
+                ASSERT( FloatUtils::isZero( costOfLastAcceptedPhasePattern ) );
+                ENGINE_LOG( "Performing local search - done" );
```
done

anwu1219 commented:
Ok, I think I've addressed all the comments.
I tested the regress1 (131 benchmarks) again with SoI on, and all tests passed. (regress1 time is < 3 minutes compared to ~8 minutes with the current master).

guykatzz reviewed (approved):
Apart from 1 small comment that's still open, LGTM
